### PR TITLE
Apply email branding for sub organizations based on the branding configured immediate parent organization

### PR DESCRIPTION
**Pull request by sadilchamishka**
### Proposed changes in this pull request

$subject

### CODE REVIEW

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java [hunk 1]
```diff
@@ -556,40 +565,83 @@ public static Notification buildNotification(Event event, Map<String, String> pl
...
+
+    /**
+     * Fetch the branding preferences from the organization hierarchy by traversing from the base org to the super org.
+     * If branding preferences enabled organization is found, that will be used as the branding preference.
+     *
+     * @param ancestorOrgIds The list of ancestor org ids including the base organization itself.
+     * @param brandingPreferenceManager The branding preference manager.
+     * @return Branding preferences of the immediate organization which has configured branding preferences.
+     */
+    private static BrandingPreference fetchBrandingPreferenceFromOrgHierarchy(List<String> ancestorOrgIds, BrandingPreferenceManager brandingPreferenceManager) {
```
Comment (inthirakumaaran) [inline components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java:611]: Check on the character length

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java [hunk 2]
```diff
@@ -556,40 +565,83 @@ public static Notification buildNotification(Event event, Map<String, String> pl
...
+                        BrandingPreferenceMgtConstants.DEFAULT_LOCALE);
+            } catch (BrandingPreferenceMgtException e) {
+                if (!BrandingPreferenceMgtConstants.ErrorMessages.ERROR_CODE_BRANDING_PREFERENCE_NOT_EXISTS.getCode()
+                        .equals(e.getErrorCode())) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Error occurred while retrieving branding preferences for organization " + tenantDomain, e);
+                    }
+                }
+            } catch (OrganizationManagementException e) {
+                log.error("Error occurred while resolving tenant from organization id: " + orgId, e);
```
Comment (inthirakumaaran) [inline components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java:634]: what is the expected default behavior here? this logic implies that if `OrganizationManagementException` occurs then we won't proceed further is that correct or are we handling this at top-level?

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java (hunk 1 above)
Comment (sadilchamishka) [reply components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java:611]: Yes. This repo doesn't enable check style. Hence lot of check style issues found. Will try to do proper styling for the new code which is going to introduced .

File: components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java [hunk 3]
```diff
@@ -177,29 +180,36 @@ public static Map<String, String> getPlaceholderValues(EmailTemplate emailTempla
...
+                BrandingPreference responseDTO;
+                if (placeHolderData.get(ORGANIZATION_ID_PLACEHOLDER) != null) {
+                    List<String> ancestorOrgIds = getOrganizationManager().getAncestorOrganizationIds(placeHolderData.get(ORGANIZATION_ID_PLACEHOLDER));
+                    responseDTO = fetchBrandingPreferenceFromOrgHierarchy(ancestorOrgIds, brandingPreferenceManager);
+                } else {
+                    responseDTO = brandingPreferenceManager.getBrandingPreference(
+                            BrandingPreferenceMgtConstants.ORGANIZATION_TYPE,
+                            placeHolderData.get("tenant-domain"),
+                            BrandingPreferenceMgtConstants.DEFAULT_LOCALE);
+                }
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java]: Yes, it worked as expected. Hence, Fixed by - https://github.com/wso2-extensions/identity-event-handler-notification/pull/188/commits/d832d14cbd5b51769250d851b32b9baaa0ab36d4

**Review by AnuradhaSK — APPROVED** [review]

**Comment by sadilchamishka** [comment]
E2E Test results
-----------------------

┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  00-pre-test/00.01-remove-existing-t      00:08        9        9        -        -        - │
  │    est-data.spec.ts                                                                            │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  00-pre-test/00.02-extract-sample-ap      00:09        2        2        -        -        - │
  │    plications.spec.ts                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  00-pre-test/00.03-disable-2fa-optio      00:05        1        1        -        -        - │
  │    ns-for-myaccount.spec.ts                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  01-business-application-management/      03:04        4        4        -        -        - │
  │    01.01_user_signin_to_configured_rea                                                         │
  │    ct_sample_application.spec.ts                                                               │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  01-business-application-management/      03:02        5        5        -        -        - │
  │    01.02_user_signin_to_configured_oid                                                         │
  │    c_sample_application.spec.ts                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  01-business-application-management/      03:06        5        5        -        -        - │
  │    01.03_user_signin_to_configured_jav                                                         │
  │    ascript_sample_application_NEW.spec                                                         │
  │    .ts                                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  01-business-application-management/      02:01        2        2        -        -        - │
  │    01.04_user_signin_to_configured_mob                                                         │
  │    ile_application.spec.ts                                                                     │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  01-business-application-management/      05:49       10       10        -        -        - │
  │    01.05_user_signin_to_configured_sam                                                         │
  │    l_sample_application.spec.ts                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  01-business-application-management/      00:03        1        1        -        -        - │
  │    01.999-post-remove-shared-resources                                                         │
  │    .spec.ts                                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  02-user-management/02.01-tenant_adm      04:24        8        8        -        -        - │
  │    in_onboards_an_application_consumer                                                         │
  │    _user_by_providing_username_and_pas                                                         │
  │    sword.spec.ts                                                                               │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  02-user-management/02.02-tenant_adm      09:55       11       11        -        -        - │
  │    in_onboards_a_new_invited_admin_use                                                         │
  │    r.spec.ts                                                                                   │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  02-user-management/02.03-tenant_adm      02:01        4        4        -        -        - │
  │    in_request_new_user_to_reset_tempor                                                         │
  │    ary_password_via_email_option.spec.                                                         │
  │    ts                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  02-user-management/02.05-tenant_adm      04:58        8        8        -        -        - │
  │    in_enable_privileged_user_mgt.spec.                                                         │
  │    ts                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  02-user-management/02.06_admin_upda      02:34        4        4        -        -        - │
  │    tes_password_validation_configurati                                                         │
  │    on_via_console_NEW.spec.ts                                                                  │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  02-user-management/02.999-post-remo      00:04        3        3        -        -        - │
  │    ve-shared-resources.spec.ts                                                                 │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.01_a      01:10        3        3        -        -        - │
  │    dmin_user_recover_password_through_                                                         │
  │    the_console.spec.ts                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.02_a      02:15        3        3        -        -        - │
  │    dmin_user_change_password_from_my_a                                                         │
  │    ccount.spec.ts                                                                              │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.03_b      00:57        2        2        -        -        - │
  │    usiness_user_update_own_profile_usi                                                         │
  │    ng_myaccount.spec.ts                                                                        │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.04_b      00:53        3        3        -        -        - │
  │    usiness_user_change_password_using_                                                         │
  │    myaccount.spec.ts                                                                           │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.05_a      02:38        3        3        -        -        - │
  │    uthorized_user_update_user_consent.                                                         │
  │    spec.ts                                                                                     │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.06_u      02:47        4        4        -        -        - │
  │    ser_self_register_in_sample_applica                                                         │
  │    tion.spec.ts                                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.07_l      05:02        2        2        -        -        - │
  │    ocal_user_signup_up_to_oidc_enterpr                                                         │
  │    ise_application.spec.ts                                                                     │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.08_f      04:51        1        1        -        -        - │
  │    edareted_user_signup_up_to_oidc_ent                                                         │
  │    erprise_application.spec.ts.spec.ts                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.09-t      04:27        4        4        -        -        - │
  │    enant_admin_enables_auto_login.spec                                                         │
  │    .ts                                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  03-self-services-management/03.10_i      01:44       10        -        1        -        9 │
  │    nvited_admin_access_asgardeo_after_                                                         │
  │    configure_backup_code_authenticator                                                         │
  │    _NEW.spec.ts                                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  03-self-services-management/03.999-      00:07        5        5        -        -        - │
  │    post-remove-shared-resources.spec.t                                                         │
  │    s                                                                                           │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  04-authentication-management/04.01_      01:37        2        2        -        -        - │
  │    business_user_authenticate_with_imp                                                         │
  │    licit_grant.spec.ts                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  04-authentication-management/04.02_      02:24        2        2        -        -        - │
  │    tenant_admin_tries_SSO_SLO_with_asg                                                         │
  │    ardeo_applications.spec.ts                                                                  │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  04-authentication-management/04.999      00:03        1        1        -        -        - │
  │    _post-remove-shared-resources.spec.                                                         │
  │    ts                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  05-multi-factor-authentication/05.0      03:47        5        5        -        -        - │
  │    1_user_attempts_to_login_to_react_a                                                         │
  │    pp_with_email_OTP_2FA_enabled.spec.                                                         │
  │    ts                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  05-multi-factor-authentication/05.0      02:47        4        4        -        -        - │
  │    2_user_attempts_to_login_to_react_a                                                         │
  │    pp_with_FIDO_enabled.spec.ts                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  05-multi-factor-authentication/05.0      02:53        5        5        -        -        - │
  │    3_user_attempts_to_login_to_react_a                                                         │
  │    pp_with_magic_link_2FA_enabled.spec                                                         │
  │    .ts                                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  05-multi-factor-authentication/05.0      01:53        8        -        1        7        - │
  │    4_user_attempts_to_login_to_react_a                                                         │
  │    pp_with_SMS_OTP_2FA_enabled_NEW.spe                                                         │
  │    c.ts                                                                                        │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  05-multi-factor-authentication/05.9      00:05        2        2        -        -        - │
  │    99-post-remove-shared-resources.spe                                                         │
  │    c.ts                                                                                        │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  06-adaptive-authentication/06.01_us      03:13        3        3        -        -        - │
  │    er_login_to_react_app_configured_ca                                                         │
  │    llchoreo_adaptive_script_with_TOTP_                                                         │
  │    2FA_enabled.spec.ts                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  06-adaptive-authentication/06.02-us      02:55        7        7        -        -        - │
  │    er_login_to_react_app_configured_ag                                                         │
  │    e_based_adaptive_authentication_ena                                                         │
  │    bled.spec.ts                                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  06-adaptive-authentication/06.03_Si      03:11        4        4        -        -        - │
  │    gn-in_to_sample_app_with_login_atte                                                         │
  │    mpt_based_adaptive_authentication_e                                                         │
  │    nabled.spec.ts                                                                              │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  06-adaptive-authentication/06.04_Si      03:46        4        4        -        -        - │
  │    gn-in_to_sample_app_with_otp_based_                                                         │
  │    login_attempt_based_adaptive_authen                                                         │
  │    tication_enabled.spec.ts                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  06-adaptive-authentication/06.05_to      04:46       11       11        -        -        - │
  │    tp_adaptive_authentication_configur                                                         │
  │    e_enroll_during_login.spec.ts                                                               │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  06-adaptive-authentication/06.06_ U      05:22        4        3        1        -        - │
  │    ser_logs_in_to_react_app_with_callC                                                         │
  │    horeo_function_in_adaptive_authenti                                                         │
  │    cation_NEW.spec.ts                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  06-adaptive-authentication/06.07_us      00:17        3        -        1        -        2 │
  │    er_login_to_react_app_configured_gr                                                         │
  │    oup_based_adaptive_authentication_N                                                         │
  │    EW.spec.ts                                                                                  │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  06-adaptive-authentication/06.999-p      00:08        2        2        -        -        - │
  │    ost-remove-shared-resources.spec.ts                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  07-federation-with-provisioning/07.      04:29        5        5        -        -        - │
  │    01_user_login_to_java_sample_applic                                                         │
  │    ation_through_oidc_enterprise_conne                                                         │
  │    ction.spec.ts                                                                               │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  07-federation-with-provisioning/07.      02:58        4        4        -        -        - │
  │    02-verify_API_response_for_configur                                                         │
  │    ed_IDP_for_MFA.spec.ts                                                                      │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  07-federation-with-provisioning/07.      00:01        2        -        -        2        - │
  │    03-user_signs_in_to_a_choreo_app_us                                                         │
  │    ing_enterprise_IDP.spec.ts                                                                  │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  07-federation-with-provisioning/07.      03:35        4        4        -        -        - │
  │    04_user_login_to_react_sample_appli                                                         │
  │    cation_using_SAML_enterprise_connec                                                         │
  │    tion.spec.ts                                                                                │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  07-federation-with-provisioning/07.      00:12        7        7        -        -        - │
  │    999-post-remove-shared-resources.sp                                                         │
  │    ec.ts                                                                                       │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  08-theme-branding/08.01_user_signin      05:32        9        9        -        -        - │
  │    g_to_branding_configured_react_samp                                                         │
  │    le_application.spec.ts                                                                      │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  08-theme-branding/08.02_user_signin      01:53        5        5        -        -        - │
  │    g_into_to_branding_configured_my_ac                                                         │
  │    count.spec.ts                                                                               │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  08-theme-branding/08.03_user_signin      06:07        3        3        -        -        - │
  │    g_to_branding_configured_b2b_sample                                                         │
  │    _application_NEW.spec.ts                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  08-theme-branding/08.999-post_remov      00:05        2        2        -        -        - │
  │    e_shared_resorces.spec.ts                                                                   │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  09-account-management/09.01-admin_c      03:12        6        6        -        -        - │
  │    onfigure_myaccount_authentication_f                                                         │
  │    low.spec.ts                                                                                 │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  09-account-management/09.02-verify_      02:19        3        3        -        -        - │
  │    a_user_can_edit_myaccount_profile.s                                                         │
  │    pec.ts                                                                                      │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  09-account-management/09.999-post_r      00:06        1        1        -        -        - │
  │    emove_shared_resorces.spec.ts                                                               │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  10-userstore-management/10.01-tenan      02:08        6        -        1        -        5 │
  │    t_admin_onboard_a_secondary_onprem_                                                         │
  │    user_store.spec.ts                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  10-userstore-management/10.02_secon      05:16        8        8        -        -        - │
  │    dory_userstore_user_signin_to_react                                                         │
  │    _sample_app_authenticate_with_2FA.s                                                         │
  │    pec.ts                                                                                      │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  10-userstore-management/10.03_secon      03:39        2        2        -        -        - │
  │    dory_userstore_user_signin_to_SAML_                                                         │
  │    sample_app.spec.ts                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  10-userstore-management/10.04-tenan      02:52        7        -        1        6        - │
  │    t_admin_onboard_a_secondary_onprem_                                                         │
  │    user_store_with_read_write_access_t                                                         │
  │    ype.NEW.spec.ts                                                                             │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  11-b2b-organization-management/11.0      08:03       13        8        1        4        - │
  │    1_business_user_login_to_b2b_app.sp                                                         │
  │    ec.ts                                                                                       │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  14-identity-providers/14.01_verify_      03:24        2        2        -        -        - │
  │    connected_apps_tab.spec.ts                                                                  │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  14-identity-providers/14.999-post-r      00:10        2        2        -        -        - │
  │    emove-shared-resources.spec.ts                                                              │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  15-scope-attribute-management/15.01      03:22        3        3        -        -        - │
  │    _validate_oidc_scope_against_access                                                         │
  │    _token_NEW.spec.ts                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  15-scope-attribute-management/15.02      03:09        3        3        -        -        - │
  │    _business_user_allow_custom_claims_                                                         │
  │    for_sample_application_NEW.spec.ts                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  15-scope-attribute-management/15.99      00:05        2        2        -        -        - │
  │    9_post-remove-shared-resources_NEW.                                                         │
  │    spec.ts                                                                                     │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  16-event-publishing/16.01_enable_ev      00:02        9        -        -        9        - │
  │    ent_publishing_console_and_confirm_                                                         │
  │    published_event_NEW.spec.ts                                                                 │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  17-user-groups-management/17.01-ten      02:46        3        3        -        -        - │
  │    ant_admin_onboard_a_new_user_group_                                                         │
  │    NEW.spec.ts                                                                                 │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  17-user-groups-management/17.999-po      00:06        1        1        -        -        - │
  │    st-remove-shared-resources.spec.ts                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  99-post-test/99.01-remove-created-t      00:15       10       10        -        -        - │
  │    est-data.spec.ts                                                                            │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  99-post-test/99.02-disable-2fa-opti      00:04        1        1        -        -        - │
  │    ons-for-myaccount.spec.ts                                                                   │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  7 of 69 failed (10%)                   2:53:45      307      256        7       28       16

**Pull request merged**